buffer/src/main/java/io/netty/buffer/ReadableObject.java

```diff
@@ -0,0 +1,130 @@
+/*
+ * Copyright 2015 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.netty.buffer;
+
+import io.netty.util.ReferenceCounted;
+
+import java.io.IOException;
+import java.nio.channels.WritableByteChannel;
+
+/**
+ * An object that contains a readable region. For simplicity, extends {@link ReferenceCounted}.
+ */
+public interface ReadableObject<T extends ReadableObject> extends ReferenceCounted {
```

Comment:
@nmittler - Have you thought about just using `ReadableObject` instead of `T` in this interface?  Or use `<T extends ReadableObject> T foo(...)` so users don't have to cast?  The usage of this generic interface seems awkward in classes that need to use `ReadableObject` in a generic way and not care about the generic argument (i.e. `AbstractNioByteChannel`).